spec/requests/omniauth_callbacks_spec.rb

```diff
@@ -0,0 +1,382 @@
+
+        user = User.find_by(email: oauth_email)
+        expect(user.token).to eq oauth_token
+      end
+
+      it 'signs in and redirect to root' do
+        get omniauth_login_omniauth_callbacks_path, params: { provider: 'google' }
+
+        expect(response).to redirect_to root_path
+      end
+
+      it 'enables oauth on user' do
+        get omniauth_login_omniauth_callbacks_path, params: { provider: 'google' }
+
+        user = User.find_by(email: oauth_email)
+        expect(user.oauth_enabled?).to eq true
+      end
+    end
+
+    context 'when google_oauth2 enabling fails' do
+      it 'redirects to sign in path' do
+        allow(User).to receive(:find_for_oauth).and_return(nil)
+
+        get omniauth_login_omniauth_callbacks_path, params: { provider: 'google' }
+
+        expect(response).to redirect_to new_user_session_path
+      end
+    end
+
+    context 'when an invitation_token is passed in' do
+      context 'when the user logs in with the same email as the invitation' do
+        it 'sets invitation accepted at on invitee' do
+          invite_user(oauth_email)
+
+          get omniauth_login_omniauth_callbacks_path, params: { provider: 'google' }
+
+          invitee = User.find_by(email: oauth_email)
+          expect(invitee.invitation_accepted_at).to be_present
+        end
+
+        it 'signs in and redirect to root' do
+          invite_user(oauth_email)
+
+          get omniauth_login_omniauth_callbacks_path, params: { provider: 'google' }
+
+          expect(response).to redirect_to root_path
+        end
+
+        it 'enables oauth on user' do
+          invite_user(oauth_email)
+
+          get omniauth_login_omniauth_callbacks_path, params: { provider: 'google' }
+
+          invitee = User.find_by(email: oauth_email)
+          expect(invitee.oauth_enabled?).to eq true
+        end
+      end
+
+      context 'when the user logs in with a different email from the invitation' do
+        let(:other_invitee_email) { 'no-invite@xyze.it' }
+
+        it 'does not invitation accepted at on invitee' do
+          invite_user(other_invitee_email)
+
+          get omniauth_login_omniauth_callbacks_path, params: { provider: 'google' }
+
+          invitee = User.find_by(email: oauth_email)
+          expect(invitee.invitation_accepted_at).to be_blank
+        end
+
+        it 'signs in and redirect to root' do
+          invite_user(other_invitee_email)
+
+          get omniauth_login_omniauth_callbacks_path, params: { provider: 'google' }
+
+          expect(response).to redirect_to root_path
+        end
+
+        it 'enables oauth on user' do
+          invite_user(other_invitee_email)
+
+          get omniauth_login_omniauth_callbacks_path, params: { provider: 'google' }
+
+          user = User.find_by(email: oauth_email)
+          expect(user.oauth_enabled?).to eq true
+        end
+      end
+    end
+
+    context 'user avatar image uploads' do
```

Comment:
```suggestion
    context 'when user avatar image uploads' do
```

Let's start context blocks off with `when`, `with`, or `without`: https://www.betterspecs.org/#contexts